[rebass] Added possibility to pass`number[]` to SpaceProps

maoueh opened the pull request:
Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Add or edit tests to reflect the change. (Run with `npm test`.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [~] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present). **I was unable to run the command `npm run lint rebass` successfully, getting `Error: Errors in typescript@next for external dependencies:
../react/index.d.ts(28,22): error TS2307: Cannot find module 'csstype'.` everytime**
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: https://rebassjs.org/props/#margin-and-padding
- [X] Increase the version number in the header if appropriate.

maoueh commented:
Ping @rhysd @ryee-dev @ jamesmckenzie

cezarneaga commented:
Wasone

maoueh commented on types/rebass/index.d.ts:
```diff
@@ -1,4 +1,4 @@
-// Type definitions for Rebass 0.3.8
+// Type definitions for Rebass 0.3.9
```
Reviewing again what I did, I think this should be `// Type definitions for Rebass 3.0.1` no?

uniqueiniquity replied:
This appears to be a malformed header. Could you remove the line in the `tslint.json` file in this folder that says `dt-header: false` and fix any errors that come up? You will want this first line to say `// Type definitions for rebass 3.0`. You are welcome to remove any other such line from the `tslint.json` file, though that might be a lot of work depending on what issues are being ignored. Thanks!

maoueh commented:
@uniqueiniquity Done, only kept 2 `tslint` directives customization and pretty sure `strict-export-declare-modifiers` could be removed by fixing up the `index.d.ts` file but I wasn't sure about the proper way.

I removed all the inner `declare` keywords` but this did not fix it. If you have more clues, I'm all in.

maoueh commented:
For `no-empty-interface`, see small custom made snippet below.

```
>  export interface HeadingProps extends TextProps {}
   type HeadingClass = React.StatelessComponent<HeadingProps>;
   export declare const Heading: HeadingClass;
```

I changed it to a `type HeadingProps = TextProps;` instead.

Also, for the declare thing, removing them all was still complaining about the lint rule due to this:

```
   export interface BoxProps extends SpaceProps<BoxClass> {
       className?: string;
       width?: number | string | ReadonlyArray<number>;
       fontSize?: number | ReadonlyArray<number>;
       css?: object;
       color?: string;
       bg?: string;
   }
>  type BoxClass = React.FunctionComponent<BoxProps>;
   export const Box: BoxClass;
```

It complains about the `BoxClass` type definition (and all others but I inlined them now to avoid the warnings).

I tried `export interface BoxProps extends SpaceProps<Box>` but the `Box` is not found with this. I'm not able to resolve this last one. I added a disable next line for now.

Check my new changeset, don't hesitate to suggest some ways to overcome the problem.

uniqueiniquity commented:
@maoueh It seems like you can just add `export` in front of `type BoxClass = React.FunctionComponent<BoxProps>;` and that will solve the lint issue.

I was poking around and trying to figure out what the type argument to `BoxProps` was doing. One thing to consider for general improvements to these types is to use `SpaceProps` from `@types/styled-system` and the types from `@types/styled-components` instead of defining your own `BaseProps` and `SpaceProps`. Just something consider; it might be worth doing in a different PR.

uniqueiniquity commented:
Makes sense to me. I think it's fine as is. Thank you for all your work on this!

Congratulations on your first DefinitelyTyped contribution!
Thank you for being a part of the community!
